core/src/impl/Kokkos_Atomic_Fetch_Sub.hpp

```diff
@@ -108,7 +112,11 @@ T atomic_fetch_sub( volatile T * const dest ,
   typename Kokkos::Impl::enable_if< sizeof(T) != sizeof(int) &&
                                     sizeof(T) == sizeof(unsigned long long int) , const T >::type val )
```

Comment:
Couldn't this one collide with the "not sizeof int and equals sizeof long" overload below, if `sizeof(long) == sizeof(unsigned long long)` (true on 64-bit Linux)?

Reply:
Yes, there are multiple redefinition problems if
- `KOKKOS_IMPL_CUDA_CLANG_WORKAROUND` is defined and 
- `KOKKOS_ENABLE_GNU_ATOMICS` or `KOKKOS_ENABLE_INTEL_ATOMICS` is defined,

compare https://godbolt.org/z/tsLuSA. I guess this should be addressed in a separate pull request if this situation can appear.

Reply:
It doesn't collide because one of them is a __device__ function and the other one is a (implicit) __host__ function. When using clang for CUDA builds __device__ and __host__ are part of the overload resolution scheme but both must be visible in both phases of the compilation (device and host) because there is presumably only one list of functions which must stay consistent. On NVCC you can't have two other wise equal overloads, and thus we use __CUDA_ARCH__ to make only one visible at a time.